tree-sitter-python/test/Test.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE TypeApplications, DisambiguateRecordFields, OverloadedStrings, TemplateHaskell #-}
+module Main (main) where
+
+import           Control.Applicative
+import           Control.Monad
+import           Data.Attoparsec.ByteString.Char8
+import           Data.ByteString (ByteString, readFile)
+import           Data.ByteString.Char8 (pack, unpack)
+import           Data.Either
+import           Data.Functor
+import           Prelude hiding (takeWhile)
+import           System.Exit (exitFailure)
+import qualified System.Path as Path
+import qualified System.Path.Directory as Path
+import           Test.Tasty
+import           Test.Tasty.HUnit
+import           TreeSitter.Python
+import qualified TreeSitter.Python.AST as Py
+import           TreeSitter.Unmarshal
+
+import qualified Manual.Examples
+
+main :: IO ()
+main = readCorpusFiles (Path.relDir "./vendor/tree-sitter-python/corpus") >>= traverse testCorpus >>= defaultMain . tests
+
+tests :: [TestTree] -> TestTree
+tests xs = testGroup "Tests"
+  [ testGroup "tree-sitter corpus tests" xs
+  , Manual.Examples.tests
+  ]
+
+testCorpus :: Path.RelFile -> IO TestTree
+testCorpus path = do
+  xs <- parseCorpusFile path
+  case xs of
+    Left e -> print ("failed to parse corpus: " <> show path) *> print e *> exitFailure
+    Right xs -> testGroup (Path.toString path) <$> traverse toPropTest xs
+  where
+    toPropTest (CorpusExample name code) = do
+      res <- parseByteString @Py.Module @() tree_sitter_python code
+      let failureMsg = "failed to parse:\n``` python\n" <> code <> "```"
+      pure $ testCase name (assertBool (unpack failureMsg) (isRight res))
+
+
+-- Read and parse tree-sitter corpus examples
+
+readCorpusFiles :: Path.RelDir ->  IO [Path.RelFile]
+readCorpusFiles dir = fmap (Path.combine dir) <$> Path.filesInDir dir
+
+data CorpusExample = CorpusExample { name :: String, code :: ByteString }
+  deriving (Eq, Show)
+
+parseCorpusFile :: Path.RelFile -> IO (Either String [CorpusExample])
+parseCorpusFile path = do
+  c <- Data.ByteString.readFile (Path.toString path)
+  pure $ parseOnly corpusParser c
+
+corpusParser :: Parser [CorpusExample]
```

Comment:
This is a pretty quick pass at parsing, feedback welcome as I'm sure there are more concise ways to express these parser combinators.

Reply:
In the next PR I'll extract this into a common testing library so that the other languages can also use this testing strategy.

Reply:
That would be great!